mixer/adapter/redisquota/redisquota.go

```diff
@@ -0,0 +1,400 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package redisquota provides a quota implementation with redis as backend.
+// The prerequisite is to have a redis server running.
+//
+//go:generate $GOPATH/src/istio.io/istio/bin/mixer_codegen.sh -f mixer/adapter/redisquota/config/config.proto
+package redisquota // import "istio.io/istio/mixer/adapter/redisquota"
+
+import (
+	"context"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/go-redis/redis"
+
+	"istio.io/istio/mixer/adapter/redisquota/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/pkg/status"
+	"istio.io/istio/mixer/template/quota"
+)
+
+var (
+	// LUA rate-limiting algorithm scripts
+	rateLimitingLUAScripts = map[config.Params_QuotaAlgorithm]string{
+		config.FIXED_WINDOW:   luaFixedWindow,
+		config.ROLLING_WINDOW: luaRollingWindow,
+	}
+)
+
+type (
+	builder struct {
+		quotaTypes    map[string]*quota.Type
+		adapterConfig *config.Params
+	}
+
+	handler struct {
+		// go-redis client
+		// connection pool with redis
+		client *redis.Client
+
+		// the limits we know about
+		limits map[string]*config.Params_Quota
+
+		// dimension hash map
+		dimensionHash map[*map[string]string]string
+
+		// list of algorithm LUA scripts
+		scripts map[config.Params_QuotaAlgorithm]*redis.Script
+
+		// indirection to support fast deterministic tests
+		getTime func() time.Time
+
+		// logger provided by the framework
+		logger adapter.Logger
+	}
+)
+
+///////////////// Configuration Methods ///////////////
+
+func (b *builder) SetQuotaTypes(quotaTypes map[string]*quota.Type) {
+	b.quotaTypes = quotaTypes
+}
+
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adapterConfig = cfg.(*config.Params)
+}
+
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+	info := GetInfo()
+
+	if len(b.adapterConfig.Quotas) == 0 {
+		ce = ce.Appendf(info.Name, "quota should not be empty")
+	}
+
+	limits := make(map[string]*config.Params_Quota, len(b.adapterConfig.Quotas))
+	for idx := range b.adapterConfig.Quotas {
+		if len(b.adapterConfig.Quotas[idx].Name) == 0 {
```

Comment:
NIT: Here and below, consider making local constant variables for expressions `b.adapterConfig.Quotas[idx]`. In the below sections `b.adapterConfig.Quotas[idx].Overrides[index]`. This will help readability, IMHO.

Reply:
Done